File: Projeto_6/Projeto6-GabrielAlves/src/professor.py
from typing import List
from usuario import UsuarioAutenticavel
from interfaces import Relatorio
from enums import TipoUsuario

class Professor(UsuarioAutenticavel, Relatorio):
    """
    Representa um professor, que pode se autenticar e gerar relatórios
    sobre as disciplinas que leciona. 
    """
    TIPO = TipoUsuario.PROFESSOR

    def __init__(self, nome: str, id_usuario: int, senha: str, departamento: str):
        super().__init__(nome, id_usuario, senha)
        self._departamento = departamento
        self._disciplinas_lecionadas: List[str] = []

    def adicionar_disciplina(self, nome_disciplina: str):
        """Adiciona uma disciplina à lista do professor."""
        self._disciplinas_lecionadas.append(nome_disciplina)

    def gerar_relatorio(self) -> None: # Implementação da interface Relatorio [cite: 12]
        """
        Gera um relatório com os detalhes do professor e as disciplinas que leciona. [cite: 12]
        """
        print("--- Relatório do Professor ---")
        print(f"Nome: {self.nome}")
        print(f"ID: {self._id_usuario}")
        print(f"Departamento: {self._departamento}")
        print("Disciplinas Lecionadas:")
        if not self._disciplinas_lecionadas:
            print("  Nenhuma disciplina atribuída.")
        else:
            for disciplina in self._disciplinas_lecionadas:
                print(f"  - {disciplina}")
        print("-" * 30)

File: Projeto 6 - Marcos Barros/Professor.cpp
#include "Professor.h"
#include <iostream>

Professor::Professor(int id, std::string nome, std::string senha, std::string departamento)
    : UsuarioAutenticavel(id, nome, TipoUsuario::PROFESSOR, senha), departamento(departamento) {}

void Professor::adicionarDisciplina(const std::string& disciplina) {
    disciplinasLecionadas.push_back(disciplina);
}

bool Professor::autenticar(const std::string& senha) const {
    std::cout << "Autenticando professor '" << this->nome << "'... ";
    return verificarSenha(senha);
}

void Professor::gerarRelatorio() const {
    std::cout << "\n--- Relatorio do Professor ---" << std::endl;
    std::cout << "ID: " << this->id << std::endl;
    std::cout << "Nome: " << this->nome << std::endl;
    std::cout << "Tipo: " << this->getTipoAsString() << std::endl;
    std::cout << "Departamento: " << this->departamento << std::endl;
    std::cout << "Disciplinas Lecionadas:" << std::endl;
    if (disciplinasLecionadas.empty()) {
        std::cout << "  Nenhuma disciplina registrada." << std::endl;
    } else {
        for (const auto& disciplina : disciplinasLecionadas) {
            std::cout << "  - " << disciplina << std::endl;
        }
    }
    std::cout << "----------------------------" << std::endl;
}